### ISSUE TITLE
Proposal: Make the system consumable function yaml easier to integrate

### ISSUE THREAD

**Issue by jacques-n** [substrait-io/substrait#65]
TLDR: @cpcloud and @jacques-n have both been working on building substrait producers and the current yaml functions are too difficult to work with. The yaml needs to be more discrete (which results in something less concise).

A key part of plan translation to/from substrait is the mapping of Substrait function definitions to producer or consumer function definitions. Currently, at least the following translators are underway:
* @jacques-n is working on converting a Calcite plan to Substrait. In Calcite, functions are mapped to very generic complex object hierarchies with embedded type resolution logic (e.g. `add(ANY, ANY)`).
* @cpcloud is working on converting Ibis plans to Substrait. In Ibis, functions are mapped to very specific function definitions `add(int32,int32)`.

This means the two efforts have touched different ends of the abstraction specification. Both efforts have found that this mapping process is overly complex using the current representation. In both cases, people have tried to create some kind additional abstractions to make the mapping process more straightforward. Since it is critical to facilitate ease of Substrait integration in many tools, this suggested that the level of abstraction/complexity is not right for easy integrations.

The current function signature yaml docs were primarily optimized to express things concisely. This includes the use of aliases, anchors and merges nodes as well as the capability to express type bounds. While the current form may be a relatively concise representation for human consumption, concision has come at the cost of integration of effort. Part of the integration complexity come from the fact that the functions are neither fully generic nor fully discrete and the current pattern for offset-based resolution. As we were discussing things, supporting fully discrete would make things easiest to reason about. 

Recommendation: Move to a representation that is much simpler. 
- Remove type bounds (only allow type concrete types or pure wildcard type)
- Remove yaml-only features

This simpler representation would require less sophistication of integrators comprehending the representation. Additionally, it also allows us to simplify the referencing scheme for binding to a specific function. Our current thoughts are that we should update the referencing scheme to express use a string pattern for referencing the specific discrete type we're bounding to. As an example, `add(i32,i32)` might become be referenced using one of `add<i32,i32>`, `add_ii`, `add:ii` or something similar to that. While a simple string representation might be somewhat, less fancy than other systems, it allows easier human consumption.

To make things less repetitive on the document side for things like description/name, we're currently thinking that we add an additional level of structure to the yaml document to share common attributes.

I'll look to post a patch proposing the changes.

**Comment by jacques-n** [comment]
@westonpace any opinions about this proposal? @cpcloud , should I move towards the actual changes? Wanted to settle on a rough outline before actually implementing.

**Comment by westonpace** [comment]
How would you encode a function that receives a list L and an integer I and returns L[I].  I believe such a function would have infinitely many discrete definitions?

**Comment by cpcloud** [comment]
that would (conceptually) be:

* `get_element(List<any>, 8)`
* `get_element(List<any>, i16)`
* `get_element(List<any>, i32)`
* `get_element(List<any>, i64)`

**Comment by jacques-n** [comment]
Actually, I was going to propose something on this and forgot to do so. I was going to suggest that we support indexed any. So: get_element(List<any1>, any1) -> any1. I don't think it creates any problems beyond what the any already produces and I think it would be used rarely. It removes the complexity of weird partial bounds, which I think is the main issue at hand.

**Comment by cpcloud** [comment]
What does the additional integer suffix mean?

**Comment by jacques-n** [comment]
They are the same type.

**Comment by westonpace** [comment]
I think I don't have enough concrete implementation experience to really say anything either way.  If I understand correctly the challenging part was:

 * Given the function name and a list of typed input expressions figure out the output type of the expression.
 
It's odd to me that the type system was the challenging part in this problem.  I would've assumed the challenging part was parsing and evaluating the output expressions.

**Comment by jacques-n** [comment]
> I would've assumed the challenging part was parsing and evaluating the output expressions.

Nobody said that wouldn't also be challenging :D

**Comment by cpcloud** [comment]
> figure out the output type of the expression
It was actually figuring out the function id for an already-typed expression that is incredibly difficult. In the ibis case we already have the output type, but the problem is to find a function id who type rules unambiguously match your output expression. This problem is very close to the problem that a type system solves and is a much more difficult problem than computing the output type of an expression.

Small correction here that hopefully makes it more clear :)

**Comment by westonpace** [comment]
Is it not:

```
for function_variant in variants:
  type_map = {}
  for arg_index, arg_type in enumerate(function_variant.arg_types):
    if arg_type.is_concrete:
      concrete_arg_type = arg_type
    else:
      if arg_type.param_name in type_map:
        concrete_arg_type = type_map[arg_type.param_name]
      else:
        concrete_arg_type = None
    if concrete_arg_type is None:
      type_map[arg_type.param_name] = incoming_func.arg_types[arg_index]
    else:
      if concrete_arg_type != incoming_func.arg_types[arg_index]:
        continue
    return function_variant
  raise Error("Could not find a function variant that matched the incoming types")
```

**Comment by westonpace** [comment]
> It was actually figuring out the function id for an already-typed expression that is incredibly difficult.

Right, but there is no reason to solve this problem unless you are trying to determine the output type of an expression.

Actually, allow me to make a (likely naive) counter-proposal.

**Motivation:** It seems that we are taking a rather onerous producer task (output type derivation) and spreading this complexity to intermediate transformation libraries as well as consumers.

**A simpler alternative is:**

 * Make the output type a required property of an expression message
 * Remove `index` from the function mapping message.
 * Keep the yaml as it is (or simplify it if you want) but now consumers and intermediate transformation libraries don't need to process it.  Even producers could ignore it if they have their own output transformation rules (e.g. Ibis, duckdb) at the risk of creating non-conforming plans (which a validation library could detect)
 
**Pros:**
 * Consumers don't need to refer to a bunch of yaml files to ensure they are giving the producer what they asked for.  They simply look at the function name, input types, and output type and decide if they can satisfy it.
 * Independent transformation libraries can now know the output type of all expressions without re-deriving it (no simple task it seems) themselves.  That makes it a lot easier to create such libraries.
  
**Cons:**
 * It is possible to create invalid plans which are in violation of functions.yaml but...
   * Is it really invalid if the producer asks for X and the consumer delivers X?
   * An independent validator library could be created which verifies the plan conforms to the yaml files it references.

**Comment by cpcloud** [comment]
@westonpace Any chance you can add some comments to that code? What is `type_map`? What is the `.param_name` field? It would make it easier to evaluate whether that does solve the problem of finding a function variant.

**Comment by cpcloud** [comment]
> It seems that we are taking a rather onerous producer task (output type derivation) and spreading this complexity to intermediate transformation libraries as well as consumers.

It's not quite that, because in the ibis case the output type is already derived. The problem statement is:

given a fully typed call, determine which overload of a particular function the call matches.

Consider an ibis expression with type `add(int64, int64) -> int64`.

For every variant I need to determine whether this signature matches the variant's type constraints.

Checking whether a signature matches in the presence of generics isn't trivial AFAICT. You need to construct a set of constraints based on the type bounds of the inputs (if there are any) and then pattern match the fully typed expression against a generic pattern with constraints. This would also have to be done for output expressions/types.

The proposal @jacques-n is suggesting doesn't require matching any generics, one only has to match signatures exactly.

> Keep the yaml as it is 

What would this mean for the different variants? Keep them? Get rid of input arguments/types and just have the name for every function?

**Comment by jacques-n** [comment]
@westonpace , the behavior that you describe can actually be done within the current pattern with the exception of the function binding. Output types for functions are in the plan. I see function semantic binding as important to ensuring consistent meaning of the plan (whether the consumer decides to respect that or not).

**Comment by cpcloud** [comment]
I think the problem arises when a producer chooses an arbitrary function id, and the consumer sees that and must reconcile that with the function YAML source of truth (if it decides to), or risk calling a function with an id that doesn't match. For example, say a producer produces `add(int32, int32) -> int32` which maps to function id 0 (because a producer didn't care look at the YAML) but the consumer knows that function id 0 maps to `add(int8, int8) -> int8`. That's I guess an error? Or, if the consumer doesn't have reference point and just executes what it's told to execute things will probably just work and function ids are an ephemeral way to prevent a bunch of strings in the proto message.

**Comment by westonpace** [comment]
The consumer isn't going to just look at the function id.  They look at the function id, the input types, and the output type.  So the consumer would see:

```
a_func_call:
  id: 0
  args:
    - ...
      output_type: int8
    - ...
      output_type: int8
  output_type: int8
```

and they would know:
```
some_function_mapping:
  - function_id: 0
  ...
    name: add
```

From this they know that the function call in question is `add(int8, int8) -> int8`.  The consumer can safely assume that the producer has already done the onerous task of validation.  They simply look in their own internal registry to decide if they have a function that can satisfy `add(int8, int8) -> int8`.

**Comment by westonpace** [comment]
> Any chance you can add some comments to that code? What is type_map? What is the .param_name field? It would make it easier to evaluate whether that does solve the problem of finding a function variant.

Basically the point is that you iterate through each variant and for each variant you have some set of parameters (T, V where V is integral).  So type map is initially:

```
T: null
V: null
```

Then you iterate through the arguments (at this point you know the concrete type of each argument in the input).  If the argument in the variant is concrete you just check to see if they match and reject if they don't.

If the argument in the variant is not concrete then you check to see if there is anything in the type map.  If there isn't you check to see if the type satisfies the constraints for the parameter and, if it doesn't, reject.  Once you determine it does satisfy the constraints you set that in the type map because that will have to be the same value throughout.

If the argument in the variant is not concrete but you already have something in the type map then you better have the exact same type here or else you reject.

If you've made it through all the input types without rejecting then you run the output derivation function on the input types to determine the output type (I agree this part is rather complex).  If the output type is anything other than "not-bindable" you've finished your work.  Otherwise you check the next variant.

It's not trivial and you'll need helper functions but it's not nearly as complex as "create a grammar, feed it to ANTLR to get a parser and then run the type derivation expression through the parser to build up an AST and then run that AST through a custom-built interpreter" which is what we are talking about for the output type derivation.

> It's not quite that, because in the ibis case the output type is already derived. 

Then in my proposal the problem is solved.  You don't need to know the variant index.

> The proposal @jacques-n is suggesting doesn't require matching any generics, one only has to match signatures exactly.

If the wildcards are indexed wildcards you still have to do some matching.  any0, any1, any2 isn't really much different than T, U, V.  Although it does remove type constraints so you don't need helper functions for "is integeral" or "is list" (although in my mind I thought we had abandoned type constraints anyways).

> What would this mean for the different variants? Keep them? Get rid of input arguments/types and just have the name for every function?

The function variants are necessary for plan validation.  In other words, someone, at some point, should probably be checking that the plan conforms to Substrait (e.g. the output types are what we would expect for all functions given their input types).  I'm just arguing that the plan validation step doesn't need to be done by everyone and could even be optional as projects are getting started.

> @westonpace , the behavior that you describe can actually be done within the current pattern with the exception of the function binding. Output types for functions are in the plan. I see function semantic binding as important to ensuring consistent meaning of the plan (whether the consumer decides to respect that or not).

You're right, output types are in the plan, I should've looked more closely.  So then the topic of this issue is "can we make it easier for plan validators to validate a plan conforms with Substrait".

**Comment by jacques-n** [comment]
> ...Although it does remove type constraints 

That is a key simplification we were trying to achieve with this proposal. All of these problems are solvable, but each additional level of complexity or abstraction in the specification adds more burden to each implementer.

> ... output derivation...

I think the common case shouldn't have systems doing output type derivation parsing and evaluation. The yaml definition is defined as a specification or contract of how a system behaves. The system (producer or consumer) doesn't have to read the specification to do what it does during actual runtime except when interacting with new unknown functions that are defined external to the system. I would expect most systems to have output type derivation rules and then declare functions associated with those rules. In many common cases, the system wouldn't have to declare new functions since they behave similar to existing specifications. 

> the topic of this issue is "can we make it easier for plan validators to validate a plan conforms with Substrait".

I wouldn't agree with this summarization. I opened the ticket because as part of implementation I was finding it difficult to have a reliable and consistent mapping between a set of definitions and an implementation that produces references to those definitions. This was partly due to the semi-abstract representation of the functions. Both myself and @cpcloud found ourselves building alternative representations to make mapping easier. The thinking here is let's actually represent the spec using something closer to the "easy to integrate representation" to make everyone have to repeat this work themselves. As I said above, this isn't about whether it is possible to build an integration with the current representation, but whether it is as easy as it could be.

Lastly, I would add that I think most systems will either: declare static mappings between internal operations and semantic definitions or do build-time dynamic binding (as opposed to doing so at runtime). The changes here also make that more straightforward, explicit and reliable.

**Comment by westonpace** [comment]
> Lastly, I would add that I think most systems will either: declare static mappings between internal operations and semantic definitions or do build-time dynamic binding (as opposed to doing so at runtime). The changes here also make that more straightforward, explicit and reliable.

This is what I'm doing and my problem is the proposed change is making this harder.  In retrospect, my problem is more the fact that I have to specify the variant index at all.  I've opened a new issue to address that.  If that could be optional then I don't care nearly as much what approach is taken here.

**Comment by westonpace** [comment]
Ok, from #69 I realized that this proposal would include dropping `Extensions.Mapping.FunctionMapping.index` which solves my concerns.

> Our current thoughts are that we should update the referencing scheme to express use a string pattern for referencing the specific discrete type we're bounding to. As an example, add(i32,i32) might become be referenced using one of add<i32,i32>, add_ii, add:ii or something similar to that. While a simple string representation might be somewhat, less fancy than other systems, it allows easier human consumption.

I think this is redundant (simple name + input types + output type should be sufficient) but not onerous if having the longer names is helpful elsewhere.  Do you want to sketch up an example reference?
